pkg/ruler/api.go

```diff
@@ -2,33 +2,47 @@ package ruler
 
 import (
 	"encoding/json"
+	"io/ioutil"
 	"net/http"
+	"net/url"
 	"sort"
 	"strconv"
 	"time"
 
 	"github.com/go-kit/kit/log"
 	"github.com/go-kit/kit/log/level"
 	"github.com/gorilla/mux"
+	"github.com/pkg/errors"
 	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
 	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/weaveworks/common/middleware"
 	"github.com/weaveworks/common/user"
+	"gopkg.in/yaml.v2"
 
 	"github.com/cortexproject/cortex/pkg/ingester/client"
+	rulefmt "github.com/cortexproject/cortex/pkg/ruler/legacy_rulefmt"
+	"github.com/cortexproject/cortex/pkg/ruler/rules"
+	store "github.com/cortexproject/cortex/pkg/ruler/rules"
 	"github.com/cortexproject/cortex/pkg/util"
 )
 
 // RegisterRoutes registers the ruler API HTTP routes with the provided Router.
-func (r *Ruler) RegisterRoutes(router *mux.Router) {
+func (r *Ruler) RegisterRoutes(router *mux.Router, middleware middleware.Interface) {
+	router = router.UseEncodedPath()
```

Comment:
This needs a comment, because it's quite hidden here, and also because later in the code we use `url.PathUnescape`. If someone accidentaly removes this line, he will break `url.PathUnescape` later in the code possibly without realizing the connection.

Handling paths with escaped characters in namespace and group also needs a unit test.

Reply:
I will add coverage for this in the integration tests. It's difficult to replicate this in a unit test since it requires the router to parse the variables.

Reply:
One possible way to test it would be to have `muxVars` global variable, that defaults to `mux.Vars` but could be set to your own function  in tests.

Reply:
One way would be injecting them into the ctx of the request that is passed to the function. But I feel like it's more important to have integration tests for this behavior since the interplay with the gorilla router is so important.

I updated the integration tests to use encoded names for the rule groups and namespaces that I think should be sufficient